src/policy_classes/ClassPolicy.hpp

```diff
@@ -0,0 +1,161 @@
+/**
+ * @file ClassPolicy.hpp
+ *
+ * @copyright Copyright (C) 2013-2014 SDML (www.srcML.org)
+ *
+ * The srcML Toolkit is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * The srcML Toolkit is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with the srcML Toolkit; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+ */
+
+#ifndef INCLUDED_CLASS_POLICY_HPP
+#define INCLUDED_CLASS_POLICY_HPP
+
+#include <srcSAXEventDispatcher.hpp>
+#include <DeclTypePolicy.hpp>
+#include <DeclDS.hpp>
+#include <FunctionSignaturePolicy.hpp>
+#include <srcSAXHandler.hpp>
+#include <exception>
+#include <stack>
+#include <list>
+
+class ClassPolicy : public srcSAXEventDispatch::EventListener, public srcSAXEventDispatch::PolicyDispatcher, public srcSAXEventDispatch::PolicyListener {
+    public:
+
+    	struct ClassData {
+    		void Clear() {
+      			className = "";
+      			methods.clear();
+      			members.clear();
+      			isStruct = false;
+   			}
+
+   			std::string className;
+   			bool isStruct = false;  //False -> Class; True -> Struct
+   			std::vector<FunctionSignaturePolicy::SignatureData> methods;
+   			std::vector<DeclData> members;
+		};
+
+        ~ClassPolicy() {
+            delete funcSigPolicy;
+            delete declTypePolicy;
+        }
+
+        ClassPolicy(std::initializer_list<srcSAXEventDispatch::PolicyListener *> listeners = {}): srcSAXEventDispatch::PolicyDispatcher(listeners) {
+            funcSigPolicy = new FunctionSignaturePolicy({this});
+            declTypePolicy = new DeclTypePolicy({this});
+
+            InitializeEventHandlers();
+        }
+
+        void Notify(const PolicyDispatcher * policy, const srcSAXEventDispatch::srcSAXEventContext & ctx) override {
+            if (typeid(FunctionSignaturePolicy) == typeid(*policy)) {
+                FunctionSignaturePolicy::SignatureData signatureData = *policy->Data<FunctionSignaturePolicy::SignatureData>();
+                data_stack.top().methods.push_back(signatureData);
+            }
+
+            else if (typeid(DeclTypePolicy) == typeid(*policy)) {
+                if(!(ctx.IsOpen(srcSAXEventDispatch::ParserState::function))) {
```

Comment:
There is a ctx.IsClosed(), I believe. That would probably be better than doing two operations here.